### PR TITLE
[MI-2705]:Added summary for status sync job

### PR DESCRIPTION
**Pull request by Kshitij-Katiyar**
https://github.com/mattermost/mattermost-plugin-mscalendar/pull/245#issuecomment-1412321999

### CODE REVIEW

File: server/jobs/status_sync_job.go [hunk 1]
```diff
@@ -23,10 +23,10 @@ func NewStatusSyncJob() RegisteredJob {
...
 	env.Logger.Debugf("User status sync job beginning")
 
-	_, err := mscalendar.New(env, "").SyncAll()
+	_, syncJobSummary, err := mscalendar.New(env, "").SyncAll()
 	if err != nil {
 		env.Logger.Errorf("Error during user status sync job. err=%v", err)
 	}
 
-	env.Logger.Debugf("User status sync job finished")
+	env.Logger.Debugf("User status sync job finished. Summary\nNumber of users processed:- %d\nNumber of users had their status changed:- %d\nNumber of users had errors:- %d", syncJobSummary.NumberOfUserProcessed, syncJobSummary.NumberOfUserStatusChange, syncJobSummary.NumberOfUserErrorInStatusChange)
```
Comment (raghavaggarwal2308) [inline server/jobs/status_sync_job.go:31]:
```suggestion
	env.Logger.Debugf("User status sync job finished.\nSummary\nNumber of users processed:- %d\nNumber of users had their status changed:- %d\nNumber of users had errors:- %d", syncJobSummary.NumberOfUserProcessed, syncJobSummary.NumberOfUserStatusChange, syncJobSummary.NumberOfUserErrorInStatusChange)
```
maybe?

File: server/mscalendar/availability.go [hunk 2]
```diff
@@ -84,24 +92,27 @@ func (m *mscalendar) syncUsers(userIndex store.UserIndex) (string, error) {
...
 	m.deliverReminders(users, calendarViews)
-	out, err := m.setUserStatuses(users, calendarViews)
+	out, numberOfUserStatusChange, numberOfUserErrorInStatusChange, err := m.setUserStatuses(users, calendarViews)
 	if err != nil {
-		return "", errors.Wrap(err, "error setting the user statuses.")
+		return "", summarySyncJob, errors.Wrap(err, "error setting the user statuses.")
 	}
 
-	return out, nil
+	summarySyncJob.NumberOfUserErrorInStatusChange = numberOfUserErrorInStatusChange
```
Comment (raghavaggarwal2308) [inline server/mscalendar/availability.go:112]:
```suggestion
	summarySyncJob.NumberOfUserErrorsInStatusChange = numberOfUserErrorsInStatusChange
```

Comment (raghavaggarwal2308) [inline server/mscalendar/availability.go:112]:
```suggestion
	summarySyncJob.NumberOfUserErrorsInStatusChange = numberOfUserErrorsInStatusChange
```

**Review by raghavaggarwal2308 — CHANGES_REQUESTED** [review]

File: server/mscalendar/availability.go [hunk 3]
```diff
@@ -26,42 +26,50 @@ const (
...
 	if err != nil {
-		return "", errors.Wrap(err, "not able to filter the super user client.")
+		return "", nil, errors.Wrap(err, "not able to filter the super user client.")
 	}
 
 	userIndex, err := m.Store.LoadUserIndex()
 	if err != nil {
 		if err.Error() == "not found" {
-			return "No users found in user index", nil
+			return "No users found in user index", nil, nil
```
Comment (ayusht2810) [inline server/mscalendar/availability.go:59]: maybe this `No user was found in the user index` or this `No users were found in the user index`?

File: server/mscalendar/availability.go [hunk 4]
```diff
@@ -26,42 +26,50 @@ const (
...
 		}
-		return "", errors.Wrap(err, "not able to load users from user index.")
+		return "", nil, errors.Wrap(err, "not able to load users from user index.")
```
Comment (ayusht2810) [inline server/mscalendar/availability.go:61]:
```suggestion
		return "", nil, errors.Wrap(err, "not able to load users from the user index.")
```

File: server/mscalendar/availability.go [hunk 5]
```diff
@@ -26,42 +26,50 @@ const (
...
 
 	return m.syncUsers(userIndex)
 }
 
-func (m *mscalendar) syncUsers(userIndex store.UserIndex) (string, error) {
+func (m *mscalendar) syncUsers(userIndex store.UserIndex) (string, *StatusSyncJobSummary, error) {
+	summarySyncJob := &StatusSyncJobSummary{}
 	if len(userIndex) == 0 {
-		return "No connected users found", nil
+		return "No connected users found", summarySyncJob, nil
```
Comment (ayusht2810) [inline server/mscalendar/availability.go:70]: same

File: server/mscalendar/availability.go [hunk 6]
```diff
@@ -26,42 +26,50 @@ const (
 	logTruncateLimit                = 5
 )
 
+type StatusSyncJobSummary struct {
+	NumberOfUserErrorInStatusChange int
+	NumberOfUserStatusChange        int
+	NumberOfUserProcessed           int
```
Comment (ayusht2810) [inline server/mscalendar/availability.go:32]:
```suggestion
	NumberOfUserErrorsInStatusChange int
	NumberOfUsersStatusChange        int
	NumberOfUsersProcessed           int
```

File: server/mscalendar/availability.go [hunk 7]
```diff
@@ -192,27 +205,32 @@ func (m *mscalendar) setUserStatuses(users []*store.User, calendarViews []*remot
...
+	return utils.JSONBlock(calendarViews), numberOfUserStatusChange, numberOfUserErrorInStatusChange, nil
 }
 
-func (m *mscalendar) setStatusFromCalendarView(user *store.User, status *model.Status, res *remote.ViewCalendarResponse) (string, error) {
+func (m *mscalendar) setStatusFromCalendarView(user *store.User, status *model.Status, res *remote.ViewCalendarResponse) (string, bool, error) {
+	isStatusChange := false
 	currentStatus := status.Status
 	if currentStatus == model.STATUS_OFFLINE && !user.Settings.GetConfirmation {
-		return "User offline and does not want status change confirmations. No status change", nil
+		return "User offline and does not want status change confirmations. No status change", isStatusChange, nil
```
Comment (ayusht2810) [inline server/mscalendar/availability.go:233]:
```suggestion
		return "User is offline and does not want status change confirmations. No status change", isStatusChange, nil
```

File: server/mscalendar/availability.go (hunk 3 above)
Comment (Kshitij-Katiyar) [reply server/mscalendar/availability.go:59]: all these messages exists before hand so i think we should not change it as they are beyond its scope

**Review by ayusht2810 — APPROVED** [review]

File: server/mscalendar/availability.go [hunk 8]
```diff
@@ -26,42 +26,50 @@ const (
 	logTruncateLimit                = 5
 )
 
+type StatusSyncJobSummary struct {
+	NumberOfUserErrorsInStatusChange int
+	NumberOfUsersStatusChange        int
```
Comment (manojmalik20) [inline server/mscalendar/availability.go:31]:
```suggestion
	NumberOfUsersFailedStatusChanged int
	NumberOfUsersStatusChanged        int
```

File: server/mscalendar/availability.go [hunk 9]
```diff
@@ -26,42 +26,50 @@ const (
...
-		return "", errors.Wrap(err, "not able to load users from user index.")
+		return "", nil, errors.Wrap(err, "not able to load the users from user index.")
 	}
 
 	return m.syncUsers(userIndex)
 }
 
-func (m *mscalendar) syncUsers(userIndex store.UserIndex) (string, error) {
+func (m *mscalendar) syncUsers(userIndex store.UserIndex) (string, *StatusSyncJobSummary, error) {
+	summarySyncJob := &StatusSyncJobSummary{}
```
Comment (manojmalik20) [inline server/mscalendar/availability.go:68]:
```suggestion
	syncJobsummary := &StatusSyncJobSummary{}
```

File: server/mscalendar/availability.go [hunk 10]
```diff
@@ -84,24 +92,27 @@ func (m *mscalendar) syncUsers(userIndex store.UserIndex) (string, error) {
...
+		return "", summarySyncJob, errors.Wrap(err, "not able to get calendar views for connected users.")
 	}
 	if len(calendarViews) == 0 {
-		return "No calendar views found", nil
+		return "No calendar views found", summarySyncJob, nil
 	}
 
 	m.deliverReminders(users, calendarViews)
-	out, err := m.setUserStatuses(users, calendarViews)
+	out, numberOfUsersStatusChange, numberOfUserErrorsInStatusChange, err := m.setUserStatuses(users, calendarViews)
```
Comment (manojmalik20) [inline server/mscalendar/availability.go:107]: Change the variable names according to the variable names inside the struct.

File: server/mscalendar/availability.go [hunk 11]
```diff
@@ -171,6 +182,7 @@ func (m *mscalendar) setUserStatuses(users []*store.User, calendarViews []*remot
 
 	var res string
 	for _, view := range calendarViews {
+		isStatusChange := false
```
Comment (manojmalik20) [inline server/mscalendar/availability.go:185]:
```suggestion
		isStatusChanged := false
```

File: server/mscalendar/availability.go [hunk 12]
```diff
@@ -65,11 +65,11 @@ func (m *mscalendar) SyncAll() (string, *StatusSyncJobSummary, error) {
 }
 
 func (m *mscalendar) syncUsers(userIndex store.UserIndex) (string, *StatusSyncJobSummary, error) {
-	summarySyncJob := &StatusSyncJobSummary{}
+	syncJobsummary := &StatusSyncJobSummary{}
```
Comment (manojmalik20) [inline server/mscalendar/availability.go:68]:
```suggestion
	syncJobSummary := &StatusSyncJobSummary{}
```

**Review by manojmalik20 — APPROVED** [review]

**Pull request merged**